File: android/src/main/java/com/dotnetdreamer/plugins/signalr/CapacitorSignalRPlugin.java
package com.dotnetdreamer.plugins.signalr;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.concurrent.CompletableFuture;

@CapacitorPlugin(name = "CapacitorSignalR")
public class CapacitorSignalRPlugin extends Plugin {

    private CapacitorSignalR implementation = new CapacitorSignalR();

    @Override
    public void load() {
        implementation.setPlugin(this);
    }

    public void notifyListenersPublic(String eventName, JSObject data) {
        notifyListeners(eventName, data);
    }

    @PluginMethod
    public void create(PluginCall call) {
        try {
            JSObject options = call.getData();
            implementation.create(options);
            
            // Start the connection
            CompletableFuture<JSObject> startFuture = implementation.start();
            
            startFuture.whenComplete((result, throwable) -> {
                if (throwable != null) {
                    call.reject("Failed to create connection: " + throwable.getMessage());
                } else {
                    call.resolve(result);
                }
            });
            
        } catch (Exception e) {
            call.reject("Error creating SignalR connection: " + e.getMessage());
        }
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        try {
            implementation.disconnect();
            call.resolve();
        } catch (Exception e) {
            call.reject("Error disconnecting: " + e.getMessage());
        }
    }

    @PluginMethod
    public void getConnectionId(PluginCall call) {
        try {
            String connectionId = implementation.getConnectionId();
            JSObject result = new JSObject();
            result.put("connectionId", connectionId);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Error getting connection ID: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void getConnectionState(PluginCall call) {
        try {
            String state = implementation.getConnectionState();
            JSObject result = new JSObject();
            result.put("state", state);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Error getting connection state: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void invoke(PluginCall call) {
        try {
            String methodName = call.getString("methodName");
            if (methodName == null) {
                call.reject("Method name is required");
                return;
            }
            
            implementation.invoke(methodName, call.getData().getJSONArray("args"));
            call.resolve();
        } catch (Exception e) {
            call.reject("Error invoking method: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void invokeWithResult(PluginCall call) {
        try {
            String methodName = call.getString("methodName");
            if (methodName == null) {
                call.reject("Method name is required");
                return;
            }
            
            CompletableFuture<Object> invokeFuture = implementation.invokeWithResult(methodName, call.getData().getJSONArray("args"));
            
            invokeFuture.whenComplete((result, throwable) -> {
                if (throwable != null) {
                    call.reject("Error invoking method: " + throwable.getMessage());
                } else {
                    JSObject response = new JSObject();
                    response.put("result", result);
                    call.resolve(response);
                }
            });
            
        } catch (Exception e) {
            call.reject("Error invoking method with result: " + e.getMessage());
        }
    }

    @PluginMethod
    public void on(PluginCall call) {
        try {
            String eventName = call.getString("eventName");
            if (eventName == null) {
                call.reject("Event name is required");
                return;
            }
            
            implementation.on(eventName);
            call.resolve();
        } catch (Exception e) {
            call.reject("Error subscribing to event: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void off(PluginCall call) {
        try {
            String eventName = call.getString("eventName");
            if (eventName == null) {
                call.reject("Event name is required");
                return;
            }
            
            implementation.off(eventName);
            call.resolve();
        } catch (Exception e) {
            call.reject("Error unsubscribing from event: " + e.getMessage(), e);
        }
    }
}

File: ios/Sources/CapacitorSignalRPlugin/CapacitorSignalRPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorSignalRPlugin)
public class CapacitorSignalRPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorSignalRPlugin"
    public let jsName = "CapacitorSignalR"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "create", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getConnectionId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getConnectionState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "invoke", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "invokeWithResult", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "on", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "off", returnType: CAPPluginReturnPromise)
    ]
    
    private let implementation = CapacitorSignalR()

    public override func load() {
        implementation.setPlugin(self)
    }

    @objc func create(_ call: CAPPluginCall) {
        do {
            let _ = try implementation.create(options: call.options as? [String: Any] ?? [:])
            
            // Start the connection
            implementation.start { [weak self] success, error, result in
                DispatchQueue.main.async {
                    if success, let result = result {
                        call.resolve(result)
                    } else {
                        call.reject(error ?? "Failed to start connection")
                    }
                }
            }
            
        } catch {
            call.reject("Error creating SignalR connection: \(error.localizedDescription)")
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        implementation.disconnect()
        call.resolve()
    }

    @objc func getConnectionId(_ call: CAPPluginCall) {
        let connectionId = implementation.getConnectionId()
        call.resolve(["connectionId": connectionId as Any])
    }

    @objc func getConnectionState(_ call: CAPPluginCall) {
        let state = implementation.getConnectionState()
        call.resolve(["state": state])
    }

    @objc func invoke(_ call: CAPPluginCall) {
        guard let methodName = call.getString("methodName") else {
            call.reject("Method name is required")
            return
        }
        
        let args = call.getArray("args", Any.self) ?? []
        
        do {
            try implementation.invoke(methodName: methodName, args: args)
            call.resolve()
        } catch {
            call.reject("Error invoking method: \(error.localizedDescription)")
        }
    }

    @objc func invokeWithResult(_ call: CAPPluginCall) {
        guard let methodName = call.getString("methodName") else {
            call.reject("Method name is required")
            return
        }
        
        let args = call.getArray("args", Any.self) ?? []
        
        implementation.invokeWithResult(methodName: methodName, args: args) { success, result, error in
            DispatchQueue.main.async {
                if success {
                    call.resolve(["result": result as Any])
                } else {
                    call.reject(error ?? "Unknown error occurred")
                }
            }
        }
    }

    @objc func on(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName") else {
            call.reject("Event name is required")
            return
        }
        
        do {
            try implementation.on(eventName: eventName)
            call.resolve()
        } catch {
            call.reject("Error subscribing to event: \(error.localizedDescription)")
        }
    }

    @objc func off(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName") else {
            call.reject("Event name is required")
            return
        }
        
        implementation.off(eventName: eventName)
        call.resolve()
    }
}
